Mantis 0006885: documentation for index operators

Octachron opened the pull request:
This pull request documents the new syntax for index operators and updates the documentation of the bigarray index operator syntax. It also mentions that it is now necessary to open the `Bigarray` module before using the `bigarray.{index}` syntax (see [0006885](http://caml.inria.fr/mantis/view.php?id=6885)). 
## TODO:
- The documentation would have to be updated again when/if [0006765](http://caml.inria.fr/mantis/view.php?id=6765) is integrated in trunk : for now, the documentation only mentions that using the `.{}` syntax without opening the `Bigarray module` is "deprecated".
## Open questions and remarks :
- The documentation have been divided in two sections, one for the custom  index operators (i.e. `.()` and `.[]` operators), and a second one for the  multidimensional index operators (i.e. the `.{}` bigarray operators). It could also make sense to fusion the two sections together.
- I have placed the two extensions at the end of the extension list in the documentation. Is this the right place?
  - Should the necessity to open the Bigarray module before using the `.{}` operators be more prominent?
- This is a first draft, remarks and criticisms are more than welcome

Octachron committed:
Mantis 0006885: documentation for index operators

This commit documents the new syntax for index operators
(.(), .[], .{}) and updates the documentation of the bigarray
specific syntax:

  * A new section "Customizable index operators" (7.28) describes the
  index operator syntax. Within this section, two subsections details
  respectively the particularities of the multidimensional index
  operator (.{}) and some potential source compatibility problem with
  the previous  bigarray specific syntax.

  * The "Syntax for bigarray access" section (7.21) has been partially
  removed and only mention that this extension has been superseded by
  the new extension and deprecated, with forward references to the new
  section 7.28 and compatibility subsection

  TODO:
  * The documentation would have to be updated again when/if the
  mantis issue 0006765 is integrated in trunk : for now, the
  documentation only mention that using the ( .{} ) syntax without
  opening the Bigarray module is "deprecated".

gasche commented:
Thanks for tackling this.

Right now you document the user-rebindable access operators in the two sections (in "Customisable index operators" you don't mention the bigarray-syntax ones, which are only in the Bigarray section). I think it would make more sense to document all of them at the same place, the "Customisable index operators" section -- because it is the same mechanism and use-cases. You could then just replace the previous bigarray-operators section by a forward-pointer to the customisable-index-operators warning and the compatibility note.

(Granted, bigarray-operators are a bit different from the other because they support the arbitrary-dimensions syntax, but I think they still make sense as a bunch.)

Octachron commented:
I fear that bringing the four variants of `.{}` in the "Customisable index operators" section might limit readability. Maybe it would be better to introduce just the two general (.{}/.{}<-) operators in this section and remarks that for the (.{}) operators, the desugaring is more complex and is explained in details in the specific subsection. Then the first subsection could focus on the notion on rebindable index operators whereas the second would explain the intricacies of the multidimensional array support. I need to think a  little more on the subject.

I like the idea of keeping a deprecated section for "Syntax for Bigarray access" with just a warning and a forward-pointer to the new section. It would help people searching in the manual why the bigarray syntax is not working anymore (or triggering a deprecation warning). Moreover, the deprecated section could eventually be removed in few versions.

gasche commented:
I think the suggestion in your first paragraph is fine.

(Your opinion is just as good as mine on this, and you are the one doing the work, so if you have an opinion on what you think is best you should go with it. Over-thinking can get in the way of getting stuff done.)

Octachron commented:
Thanks for your comment. The problem with writing the documentation after the code is that I don't have a very fresh point of view on the subject; your advice is therefore quite valuable. 

I have updated the documentation accordingly to my last comment:
- The two sections "Customisable index operators" and  "Syntax for multidimensional array access" have been reorganised:
  - The section "Syntax for multidimensional array access" is now a subsection entitled "Multidimensional index operators"
  - The (.{}) index operator family is introduced in the first section with a forward reference to the multidimensional subsection
  - I have added a non-bigarray example in the multidimensional subsection
- I have added a "Compatibility with bigarray library" subsection to the "Customisable index operators" section after the multidimensional subsection
- The "Syntax for bigarray access" section have been reintroduced to mention that this extension has been superseded by the new extension and deprecated, with forward references to the new section and compatibility subsection

I will squash the two commits after I stop fiddling with the documentation.

gasche commented:
I like the last version. Do tell me when you're done fiddling, I'll merge at some point.

Octachron commented:
I have finally found the time to reread the patch properly and corrected some minors erros. As far as I am concerned, you can merge whenever you want.

gasche commented:
Thanks! I merged this in the manual repository.
